### PR TITLE
✨ [Story video] Do not downgrade if loaded

### PR DESCRIPTION
**Pull request by mszylkowski**
After: #32618

Closes: #29498

If a video is more than 80% loaded, do not downgrade (because that will force re-downloading unnecessarily).
Also, if a video is in the preload stage (because of distance or similar) do not call load.

To get the percentage of the video buffered, we use:

```js
function getBufferedPercentage(videoEl) {
  let bufferedSum = 0;
  for (let i = 0; i < videoEl.buffered.length; i++) {
    bufferedSum += videoEl.buffered.end(i) - videoEl.buffered.start(i);
  }
  return bufferedSum / videoEl.duration;
}
```

and then when downgrading the videos, do `if (getBufferedPercentage(videoEl) < THRESHOLD) {/* downgrade code */}` where the thdeshold is 80% of the video loaded.

### CODE REVIEW

File: extensions/amp-video/0.1/flexible-bitrate.js [hunk 1]
```diff
@@ -289,3 +301,16 @@ function currentSource(video) {
...
+ * @private
+ * @param {!Element} videoEl
+ * @return {number}
+ */
+function getBufferedPercentage(videoEl) {
+  let bufferedSum = 0;
+  for (let i = 0; i < videoEl.buffered.length; i++) {
+    bufferedSum += videoEl.buffered.end(i) - videoEl.buffered.start(i);
+  }
+  return bufferedSum / videoEl.duration;
```
Comment (gmajoulet) [inline extensions/amp-video/0.1/flexible-bitrate.js:315]: You should add an early return if readyState is 0, as `duration` would not be available

File: extensions/amp-video/0.1/flexible-bitrate.js [hunk 2]
```diff
@@ -234,11 +240,17 @@ export class BitrateManager {
...
-      if (!video.paused) {
+      if (
+        !video.paused ||
+        getBufferedPercentage(video) > BUFFERED_THRESHOLD_PERCENTAGE
+      ) {
         return;
       }
       this.sortSources_(video);
-      video.load();
+      // If video has metadata, load again. Avoid loading videos that are far or were unloaded.
```
Comment (gmajoulet) [inline extensions/amp-video/0.1/flexible-bitrate.js:250]: We'll have to spend time testing this but from what I know, not calling `load()` will not restart the media selection process, and so even if you sort the sources, it won't affect which source plays on `play()`.

tldr: you need to call load() at some point, if it becomes too tricky let's save this optimization for a follow up PR with dedicated testing

File: extensions/amp-video/0.1/flexible-bitrate.js (hunk 1 above)
Comment (mszylkowski) [reply extensions/amp-video/0.1/flexible-bitrate.js:316]: SG, I think it's best to directly check for the duration to catch edge cases like `duration = 0` as well as `duration = null` by doing

```js
if (!videoEl.duration) {
  return 0;
}
```
at the start of the function

File: extensions/amp-video/0.1/flexible-bitrate.js (hunk 2 above)
Comment (mszylkowski) [reply extensions/amp-video/0.1/flexible-bitrate.js:251]: Sounds good. I think this optimization would only be useful if the mediapool replaces a video but doesn't call load (so it will call it later), but it shouldn't get triggered with today's code anyways

File: extensions/amp-video/0.1/flexible-bitrate.js (hunk 1 above)
Comment (gmajoulet) [reply extensions/amp-video/0.1/flexible-bitrate.js:316]: Okay that's a good point. Please add a code comment to say that `duration` could be `NaN` (not available yet!) or `0`. `NaN` is always tricky to handle

File: extensions/amp-video/0.1/test/test-flexible-bitrate.js [hunk 3]
```diff
@@ -78,6 +78,21 @@ describes.fakeWin('amp-video flexible-bitrate', {}, (env) => {
       expect(currentBitrates(v1)[0]).to.equal(1000);
       expect(v1.currentSrc).to.equal('http://localhost:9876/1000.mp4');
     });
+
+    it('should not lower bitrate on loaded video', () => {
+      const m = getManager('4g');
+      const v0 = getVideo([4000, 1000, 3000, 2000]);
+      env.sandbox.stub(v0, 'duration').value(0);
```
Comment (gmajoulet) [inline extensions/amp-video/0.1/test/test-flexible-bitrate.js:85]: Can you add one more test that doesn't have a duration of 0?

File: extensions/amp-video/0.1/flexible-bitrate.js [hunk 4]
```diff
@@ -35,6 +35,9 @@ const BITRATE_BY_EFFECTIVE_TYPE = {
   '5g': 5000,
 };
 
+/** @const {number} The percentage that videos need to be buffered to not downgrade them */
```
Comment (gmajoulet) [inline extensions/amp-video/0.1/flexible-bitrate.js:38]: I think this comment could be a bit easier to understand with the important information at the beginning: "Do not downgrade the quality of a video that has loaded enough content"

File: extensions/amp-video/0.1/flexible-bitrate.js [hunk 5]
```diff
@@ -293,3 +299,20 @@ function currentSource(video) {
     })
   );
 }
+
+/**
+ * @private
+ * @param {!Element} videoEl
+ * @return {number}
```
Comment (gmajoulet) [inline extensions/amp-video/0.1/flexible-bitrate.js:306]: Please specify that it returns `[0, 1]` because with the `percentage` keyword it could also be `[0, 100]`

File: extensions/amp-video/0.1/test/test-flexible-bitrate.js (hunk 3 above)
Comment (mszylkowski) [reply extensions/amp-video/0.1/test/test-flexible-bitrate.js:85]: This is definitely a typo, this is supposed to be 10 so the buffer is sufficiently filled 😁

**Review by gmajoulet — APPROVED** [review]
I'm surprised you didn't use `.reduce()` here but I'll take it haha 👀

**Comment by mszylkowski** [comment]
Lol the thing is that buffered is not really an array, so `.reduce` would require some sort middlestep that doesn't make sense to overcomplicate in this case
